src/blockchain/src/notifier.rs

```diff
@@ -0,0 +1,152 @@
+use std::fmt;
+use std::fmt::{Debug, Formatter};
+
+use crate::block::{Block, Transaction};
+use crate::blockchain::Blockchain;
+
+#[derive(Hash, Eq, PartialEq, Clone)]
+pub struct BlockchainError;
+
+// But we require certain bounds to get things done...
+impl Blockchain {
+    // should we borrow or own this transaction?
+    pub fn submit_transaction<CallBack: 'static + FnOnce(Transaction)>(
+        &mut self,
+        trans: Transaction,
+        on_done: CallBack,
+    ) -> &mut Self {
+        self.trans_observers.insert(trans, Box::new(on_done));
+        self
+    }
+
+    pub fn notify_transaction_settled(&mut self, trans: Transaction) {
+        // if there were no observers, we don't care
+        if let Some(on_settled) = self.trans_observers.remove(&trans) {
+            on_settled(trans)
+        }
+    }
+
+    pub fn add_block_listener<CallBack: 'static + FnMut(Block)>(
+        &mut self,
+        on_block: CallBack,
+    ) -> &mut Self {
+        self.block_observers.push(Box::new(on_block));
+        self
+    }
+
+    pub fn notify_block_event(&mut self, block: Block) -> &mut Self {
+        self.block_observers
+            .iter_mut()
+            .for_each(|notify| notify(block.clone()));
+        self
+    }
+}
+
+impl fmt::Display for BlockchainError {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "invalid first item to double")
+    }
+}
+
+impl Debug for Blockchain {
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        let Blockchain {
+            trans_observers: _,
+            genesis_block,
+            blocks,
+            block_observers: _,
+        } = self;
+
+        f.debug_struct("Blockchain")
```

Comment:
Maybe print the number of observers? give some clue

Reply:
Fair.